### PR TITLE
fix: Issue where CodeAct agent was trying to log cost on local llm and throwing Undefined Model execption out of litellm

### PR DESCRIPTION
**Pull request by bartshappee**
Fix this error introduced in  https://github.com/OpenDevin/OpenDevin/pull/1612 when using local model.

image

@rbren pls review this pr, thanks.

### CODE REVIEW

**Review by rbren — COMMENTED** [review]
Thanks for the fix!

Instead of this implementation, what if we changed the `completion_cost` function in llm.py to return `0.00` when working with local models?

**Comment by bartshappee** [comment]
> Thanks for the fix!
> 
> Instead of this implementation, what if we changed the `completion_cost` function in llm.py to return `0.00` when working with local models?

Is there a better way than checking that `'localhost' not in self.base_url:` to know if the model is local?

**Comment by rbren** [comment]
Ah I didn't realize we were just exposing the `completion_cost` function directly from LiteLLM.

What I'd suggest is adding a new function to the LLM class in llm.py:
```
def completion_cost(self, prompt):
  try:
    return litellm.completion_cost(prompt)
  except Exception:
    return 0
```

**Comment by bartshappee** [comment]
I added the localhost check so that there was at least a warning in the log, if not a local model, so that we aren't logging $0 even if there is a remote cost but an unsupported model in litellm's costing module

**Review by rbren — APPROVED** [review]
Thanks for the fix!

**Review by rbren — CHANGES_REQUESTED** [review]
Looks like there's a slight issue when base_url isn't set
Screenshot 2024-05-09 at 10 13 59 PM

File: agenthub/codeact_agent/codeact_agent.py [hunk 1]
```diff
@@ -265,3 +263,12 @@ def step(self, state: State) -> Action:
 
     def search_memory(self, query: str) -> List[str]:
         raise NotImplementedError('Implement this abstract method')
+
+    def logCost(self, response):
```
Comment (SmartManoj) [inline agenthub/codeact_agent/codeact_agent.py:267]: not snake_case. Also, why are you moving this to a function?

File: opendevin/llm/llm.py [hunk 2]
```diff
@@ -178,6 +178,29 @@ def get_token_count(self, messages):
...
+            self.base_url is not None
+            and 'localhost' not in self.base_url
+            and '127.0.0.1' not in self.base_url
+            and '0.0.0.0' not in self.base_url
+        ):
+            try:
+                cost = litellm_completion_cost(completion_response=response)
+                return cost
+            except Exception:
+                logger.warning('Cost calculation not supported for this model.')
```
Comment (SmartManoj) [inline opendevin/llm/llm.py:201]: We should simply ignore the exception if `if 'Model not in model_prices_and_context_window. json' in str(e)`  as we are running local models to not worry about the cost itself else there would be huge warnings.

**Review by SmartManoj — CHANGES_REQUESTED** [review]

File: agenthub/codeact_agent/codeact_agent.py [hunk 3]
```diff
@@ -265,3 +263,12 @@ def step(self, state: State) -> Action:
...
+        cur_cost = self.llm.completion_cost(response)
+        self.cost_accumulator += cur_cost
+        logger.info(
+            'Cost: %.2f USD | Accumulated Cost: %.2f USD',
+            cur_cost,
+            self.cost_accumulator,
+        )
```
Comment (SmartManoj) [inline agenthub/codeact_agent/codeact_agent.py:274]: not snake_case. Also, why are you moving this to a function?

Comment (SmartManoj) [inline agenthub/codeact_agent/codeact_agent.py:274]: not snake_case. Also, why are you moving this to a function?

**Review by SmartManoj — CHANGES_REQUESTED** [review]

File: agenthub/codeact_agent/codeact_agent.py (hunk 3 above)
Comment (bartshappee) [reply agenthub/codeact_agent/codeact_agent.py:274]: Thanks updating naming convention.  Using a function here is a cleaner implementation, you don't need to clutter the main function with 6/7 lines of simple logging code

File: opendevin/llm/llm.py (hunk 2 above)
Comment (bartshappee) [reply opendevin/llm/llm.py:201]: I've reached out to the LiteLLM team on how to handle this correctly.  The whole point of this code is to not log a warning when running local models, only when running remote unknown (from a pricing perspective) models

Comment (SmartManoj) [reply opendevin/llm/llm.py:201]: Got it. then we should use a config or environment variable for the time being. else the user may skip the warning

File: opendevin/llm/llm.py [hunk 4]
```diff
@@ -207,6 +207,25 @@ def get_token_count(self, messages):
...
+            boolean: True if executing a local model.
+        """
+        if self.base_url is not None:
+            if (
+                'localhost' not in self.base_url
+                and '127.0.0.1' not in self.base_url
+                and '0.0.0.0' not in self.base_url
+            ):
+                return True
+        elif self.model_name is not None:
```
Comment (enyst) [inline opendevin/llm/llm.py:224]: Is it possible to run a local model and _not_ set base_url?

Comment (SmartManoj) [reply opendevin/llm/llm.py:224]: `api_base` is passed
https://litellm.vercel.app/docs/providers/ollama
https://litellm.vercel.app/docs/providers/openai_compatible

Comment (bartshappee) [reply opendevin/llm/llm.py:224]: api_base is not required, I tested with both ollama and openai model names and if the base_url is not provided it defaults to localhost and openai respectively

File: opendevin/llm/llm.py (hunk 2 above)
Comment (bartshappee) [reply opendevin/llm/llm.py:244]: The code change that broke local models is already going to introduce 1 line of log per step, this just puts a warning before that line, not drastically impacting logging

**Review by rbren — APPROVED** [review]

**Comment by rbren** [comment]
Thanks for this!

**Comment by simplesisu** [comment]
Error is still there unfortunately, atleast for me:
```
21:55:24 - opendevin:ERROR: agent_controller.py:150 - Error in loop
Traceback (most recent call last):
  File "/app/opendevin/controller/agent_controller.py", line 145, in _run
    finished = await self.step(i)
               ^^^^^^^^^^^^^^^^^^
  File "/app/opendevin/controller/agent_controller.py", line 261, in step
    action = self.agent.step(self.state)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/agenthub/codeact_agent/codeact_agent.py", line 231, in step
    cur_cost = completion_cost(completion_response=response)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/litellm/utils.py", line 4453, in completion_cost
    raise e
  File "/app/.venv/lib/python3.12/site-packages/litellm/utils.py", line 4439, in completion_cost
    ) = cost_per_token(
        ^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/litellm/utils.py", line 4254, in cost_per_token
    raise litellm.exceptions.NotFoundError(  # type: ignore
litellm.exceptions.NotFoundError: Model not in model_prices_and_context_window.json. You passed model=ollama/llama3:8b-instruct-q8_0. Register pricing for model - https://docs.litellm.ai/docs/proxy/custom_pricing

21:55:24 - opendevin:INFO: agent_controller.py:193 - Setting agent state from AgentState.RUNNING to AgentState.ERROR

I am running latest commit

```

**Comment by SmartManoj** [comment]
@simplesisu Which commit?

> File "/app/agenthub/codeact_agent/codeact_agent.py", line 231, in step

This line is changed in this PR.

**Pull request merged**